### PR TITLE
[tune/execution] Update staged resources in a fixed counter for faster lookup

### LINKED ISSUE #32077: [tune] Possible performance regression in test_result_throughput_cluster
### What happened + What you expected to happen

See #31337 - test_result_throughput_cluster takes 135 seconds to run. Previously, 120 seconds was easily enough.

We should investigate the cause for this regression.

### Versions / Dependencies

master

### Reproduction script

-

### Issue Severity

High: It blocks me from completing my task.

### PR DESCRIPTION
**Pull request by krfricke**
Signed-off-by: Kai Fricke <kai@anyscale.com>

## Why are these changes needed?

In #30016 we migrated Ray Tune to use a new resource management interface. In the same PR, we simplified the resource consolidation logic. This lead to a performance regression first identified in #31337.

After manual profiling, the regression seems to come from `RayTrialExecutor._count_staged_resources`. We have 1000 staged trials, and this function is called on every step, executing a linear scan through all trials.

This PR fixes this performance bottleneck by keeping state of the resource counter instead of dynamically recreating it every time. This is simple as we can just add/subtract the resources whenever we add/remove from the `RayTrialExecutor._staged_trials` set.

Manual testing confirmed this improves the runtime of `tune_scalability_result_throughput_cluster` from ~132 seconds to ~122 seconds, bringing it back to the same level as before the refactor.

## Related issue number

Closes #32077

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: release/ray_release/alerts/tune_tests.py [hunk 1]
```diff
@@ -44,7 +44,7 @@ def handle_result(
         target_time = 900 if not was_smoke_test else 400
     elif test_name == "result_throughput_cluster":
         target_terminated = 1000
-        target_time = 160
+        target_time = 130
```
Comment (cadedaniel) [inline release/ray_release/alerts/tune_tests.py:47]: I changed a few more `130`'s than this one https://github.com/ray-project/ray/pull/32062/files

File: python/ray/tune/execution/ray_trial_executor.py [hunk 2]
```diff
@@ -233,6 +233,7 @@ def __init__(
 
         # Trials for which we requested resources
         self._staged_trials = set()
+        self._staged_resources = Counter()
```
Comment (xwjiang2010) [inline python/ray/tune/execution/ray_trial_executor.py:236]: Some comment on what the attribute is for and why?

File: release/ray_release/alerts/tune_tests.py (hunk 1 above)
Comment (krfricke) [reply release/ray_release/alerts/tune_tests.py:47]: Thanks, updated! Running both tests now

**Review by rkooo567 — APPROVED** [review]

**Comment by krfricke** [comment]
Release test passes: https://buildkite.com/ray-project/release-tests-pr/builds/27037#018608cc-d313-4c32-9369-424fd498f0b3

**Review by xwjiang2010 — APPROVED** [review]

**Pull request merged**